### PR TITLE
Tila 2853: add search filters

### PR DESCRIPTION
**Pull request by vincit-matu**
Adds filters for date, time, duration and equipment in single search

### CODE REVIEW

File: apps/ui/components/single-search/SearchForm.tsx [hunk 1]
```diff
@@ -105,71 +114,126 @@ const StyledSelect = styled(Select<OptionType>)`
...
     & > div {
       border-right-width: 0;
     }
+  }
+`;
 
-    & + .inputGroupEnd > div {
-      border-left-width: 2px;
+const OptionalFilters = styled(ShowAllContainer)<{
+  children: ReactNode[];
```
Comment (joonatank) [inline apps/ui/components/single-search/SearchForm.tsx:138]: Is the children used still?

File: apps/ui/components/single-search/SearchForm.tsx [hunk 2]
```diff
@@ -105,71 +114,126 @@ const StyledSelect = styled(Select<OptionType>)`
...
-    label {
-      width: calc(${({ children }) => children.length} * 100%);
-    }
-  }
-
-  .inputGroupEnd {
-    & > div {
-      border-left-width: 0;
+const StyledCheckBox = styled(Checkbox)`
+  &&&& {
```
Comment (joonatank) [inline apps/ui/components/single-search/SearchForm.tsx:118]: Is this a thing? documentation has only `&` and `&&`, I think `&&&` works probably by combining the other variants.

File: apps/ui/components/single-search/SearchForm.tsx [hunk 3]
```diff
@@ -86,15 +87,23 @@ const Filters = styled.div<{ $areFiltersVisible: boolean }>`
...
   }
 
   @media (min-width: ${breakpoints.l}) {
-    grid-template-columns: 1fr 1fr 1fr;
+    grid-template-columns: repeat(3, 1fr);
   }
 `;
 
-const StyledSelect = styled(Select<OptionType>)`
+const Filters = styled.div<{ children: ReactNode[] }>`
```
Comment (joonatank) [inline apps/ui/components/single-search/SearchForm.tsx:102]: Where is the `children` used?

File: apps/ui/components/single-search/FilterTagList.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                !Number.isNaN(Number(formValues.duration))
+                  ? durationTranslation(Number(formValues.duration))
+                  : label}
+              </StyledTag>
+            );
+          })}
+      </FilterTags>
+      {formValueKeys.length > 0 && (
+        <ResetButton
+          aria-label={t("searchForm:resetForm")}
```
Comment (joonatank) [inline apps/ui/components/single-search/FilterTagList.tsx:123]: aria-label is unnecessary if the inner html is same as the label

File: apps/ui/components/single-search/SearchForm.tsx [hunk 5]
```diff
@@ -86,15 +87,23 @@ const Filters = styled.div<{ $areFiltersVisible: boolean }>`
...
+  ${filterGrid}
+`;
+
+const StyledSelect = styled(Select<OptionType>)<{ name?: string }>`
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:106]: Name is only used in the `<StyledSelect>` element. Using a `name` attribute in the element results in errors, as `StyledSelect` doesn't otherwise expect it.

File: apps/ui/components/single-search/SearchForm.tsx [hunk 6]
```diff
@@ -179,286 +243,401 @@ const SubmitButton = styled(MediumButton)`
...
+              />
+
+              <StyledSelect
+                id="participantMaxCountFilter"
+                name="maxPersons"
+                placeholder={t("common:maximum")}
+                options={[emptyOption(t("common:maximum"))].concat(
+                  participantCountOptions
+                )}
+                label="&nbsp;"
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:560]: MinPersons and maxPersons are part of a "combo input" (like the date and time inputs too). The HDS Select expects a label attribute for each instance of `<Select>`, but min/maxPersons share a single label. Thus the second label is a non-breaking space. Could just as well use `" "`, but using an empty string breaks the styling (as then the `<label>` is not rendered at all - resulting in the two dropdowns not being aligned to each other.
Tl;dr: Hack to get around HDS limitations.

File: apps/ui/components/single-search/SearchForm.tsx [hunk 7]
```diff
@@ -179,286 +243,401 @@ const SubmitButton = styled(MediumButton)`
...
+              endDate={fromUIDate(String(getValues("dateEnd")))}
+              onChangeStartDate={(date: Date | null) =>
+                setValue("dateBegin", date != null ? toUIDate(date) : "")
+              }
+              onChangeEndDate={(date: Date | null) =>
+                setValue("dateEnd", date != null ? toUIDate(date) : "")
+              }
+              labels={{
+                begin: t("searchForm:dateFilter"),
+                end: t(" "),
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:472]: Ah, the empty string is not supposed to be within a `t()`. But here too, it's the same thing as with the `&nbsp;` earlier. I'll replace `t(" ")` with `&nbsp;`.

File: apps/ui/components/single-search/SearchForm.tsx (hunk 2 above)
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:118]: It's a result of some frustration regarding getting the `<Checkbox>` in question to follow the style declarations in that styled-component. `&` and `&&` doesn't do the trick - I checked. However apparently `&&&` is enough - so I removed one of those. 😅
(However it does appear to be a thing - as it also worked with `&&&&`. Could be that it combines two `&&`, dunno..)

File: apps/ui/components/single-search/SearchForm.tsx (hunk 3 above)
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:102]: Nowhere anymore. It's a relic from the former Filters `grid`, which was constructed dynamically based on `children.length`. Now it's removed. 👍

File: apps/ui/components/single-search/FilterTagList.tsx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+`;
+
+const FilterTags = styled.div`
+  display: flex;
+  flex-wrap: wrap;
+  gap: var(--spacing-s);
+  margin-right: var(--spacing-m);
+`;
+
+const ResetButton = styled(StyledTag).attrs({
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/FilterTagList.tsx:61]: ...but it's using a HDS `<Tag>` element, not a `div`. Tag has a `onDelete` function feature: "Callback function fired when the delete icon is clicked. If set, a delete button will be shown."

File: apps/ui/components/single-search/FilterTagList.tsx (hunk 4 above)
Comment (vincit-matu) [reply apps/ui/components/single-search/FilterTagList.tsx]: Okiedokie - will remove it. This is how it was done before, I simply refactored it from `SearchForm` to it's own component for clarity, as `SearchForm` started to be really stuffed and hard to read.

File: packages/common/src/components/form/TimeRangePicker.tsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+      });
+      minute += interval;
+      // Reset the minute counter, and increment the hour counter if necessary
+      if (minute >= 60) {
+        minute = 0;
+        hour += 1;
+      }
+    }
+    // we need to add the minute times to the beginning of the duration options
+    return times as OptionType[];
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply packages/common/src/components/form/TimeRangePicker.tsx:114]: No, probably not. I think I was first casting it, before specifying it in the function declaration and this was left behind (..and didn't give any errors 😇). I'll remove the extraneous typecasting - but just out of curiosity, is there some harm in typecasting a variable to the same type that it infers?

File: apps/ui/components/single-search/FilterTagList.tsx [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+const StyledTag = styled(Tag)`
+  font-size: var(--fontsize-body-m);
+`;
+
+const FilterTags = styled.div`
+  display: flex;
+  flex-wrap: wrap;
+  gap: var(--spacing-s);
+  margin-right: var(--spacing-m);
+`;
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/FilterTagList.tsx]: They are just refactored to their own element, and the styles were moved pretty much ad verbatim to the `FilterTagList.tsx` component. They could probably be `ul` and `li`, but didn't just come to think of it as they seemed to work fine as they were. Also, the "list items" in this case aren't `div`s, but `Tag`s.
I guess I could have `FilterTags` as an `ul`, but then the HDS `Tag`s would need to be wrapped in `li` elements, and the whole shebang would need restyling. Do you think it's necessary? Another option would also be to just call the component `FilterTags`, but as discussed before I don't like component names in plural (== "FilterTag" and "FilterTags" are easy to confuse with each other - hence "FilterTagList").
Also, `TagControls` is removed now. It was unnecessary as the tags are moved inside the `BottomContainer` in `SearchForm`, to get them aligned with the search button. So the result is `FilterTags` being a styled `div`, which wraps the `StyledTag`s (== HDS Tags).

File: apps/ui/components/single-search/SearchForm.tsx [hunk 11]
```diff
@@ -179,286 +243,401 @@ const SubmitButton = styled(MediumButton)`
...
+      });
+      minute += 30;
+      // Reset the minute counter, and increment the hour counter if necessary
+      if (minute === 60) {
+        minute = 0;
+        hour++;
+      }
     }
+    // we need to add the minute times to the beginning of the duration options
+    return durationMinuteOptions().concat(...times) as OptionType[];
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:328]: Yeah, same thing as with the other typecast in the return statement - left over from before I specified the type in the function declaration. And you're right with the spread - weirdly it seems to work fine even with the spread. 😮

File: apps/ui/components/single-search/SearchForm.tsx (hunk 1 above)
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:138]: Nope, same as with the other `children` - a relic from a styled-component with a dynamic grid. But I'd rather specify the styles just written out and specify all needed styles myself. We know how many children the `Filters` & `OptionalFilters` have - so why be unnecessarily clever. :P
This is removed now.

File: apps/ui/components/single-search/SearchForm.tsx [hunk 12]
```diff
@@ -105,71 +114,126 @@ const StyledSelect = styled(Select<OptionType>)`
...
+      &:empty {
+        row-gap: 0;
+        ~ [class*="ShowAllContainer__ToggleButtonContainer"] {
+          margin-top: 0;
+        }
+      }
+    }
+    > [class*="ShowAllContainer__ToggleButtonContainer"] {
+      margin-top: var(--spacing-s);
+    }
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:161]: Not sure what you're referring to here, but I added comments about the usage of the `&:empty` part - as it's the only part that could be a bit cryptic. Basically what's happening here, is that we need to remove the `row-gap` of the grid within `ShowAllContainer` and the `margin-top` of `ShowAllContainer__ToggleButtonContainer` when the contents are hidden (== `ShowAllContainer` is :empty).

File: apps/ui/pages/search/single.tsx [hunk 13]
```diff
@@ -276,16 +308,21 @@ const SearchSingle = (): JSX.Element => {
...
-                <ReservationUnitCard reservationUnit={ru} key={ru.id} />
-              ))}
+              items={reservationUnits?.map((ru) =>
+                ru ? (
+                  <ReservationUnitCard reservationUnit={ru} key={ru.id} />
+                ) : (
+                  // eslint-disable-next-line react/jsx-no-useless-fragment
+                  <></> // items attribute expects an array of JSX elements
+                )
+              )}
```
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/pages/search/single.tsx]: Nope. 😅

File: apps/ui/components/single-search/SearchForm.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx]: I also added a possibility to pass `ariaBegin` and `ariaEnd` inside the `labels` object. Along with the same fallbacks as for the actual labels (dateSelector:labelBeginDate & dateSelector:labelEndDate).

File: packages/common/src/components/form/TimeRangePicker.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (joonatank) [reply packages/common/src/components/form/TimeRangePicker.tsx]: Type casts remove type checks, technically TS 5+ doesn't allow casting unmatched types without going through unknown but it might (depends on settings and types).
So `undefined as number` might be valid code and it's gonna break on runtime. In this case if the array might be undefined the cast would probably be valid, but would crash on runtime.

File: apps/ui/components/single-search/SearchForm.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (joonatank) [reply apps/ui/components/single-search/SearchForm.tsx:535]: There is no `aria-labelledby`? I guess it's fine.

File: apps/ui/components/single-search/SearchForm.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (joonatank) [reply apps/ui/components/single-search/SearchForm.tsx:85]: It shouldn't require it. If `name` is defined for HDS Select the prop should be passed through when inheriting with styled. To me it seems like `name` is not defined for Select (and it's not), adding it just hides the typescript error without doing anything.

[the remark this answers is not in the document]
Comment (vincit-matu) [reply apps/ui/components/single-search/SearchForm.tsx:85]: Yes... "Doing this properly would be to code a common Dropdown-component of our own from scratch, which wraps a HDS Select to use it properly with react-hook-form (like I did with TextInput)."

**Review by joonatank — APPROVED** [review]
LGTM
